sources/packages/backend/apps/workers/src/services/disbursement-schedule/disbursement-schedule.service.ts

```diff
@@ -0,0 +1,179 @@
+import { Injectable } from "@nestjs/common";
+import {
+  ApplicationStatus,
+  DisbursementSchedule,
+  MSFAANumber,
+  OfferingIntensity,
+  RecordDataModelService,
+} from "@sims/sims-db";
+import { CustomNamedError } from "@sims/utilities";
+
+import { DataSource, IsNull, Not } from "typeorm";
+import { MSFAANumberService } from "..";
+import {
+  APPLICATION_MSFAA_ALREADY_ASSOCIATED,
+  DISBURSEMENT_NOT_FOUND,
+} from "../../constants";
+
+@Injectable()
+export class DisbursementScheduleService extends RecordDataModelService<DisbursementSchedule> {
+  constructor(
+    dataSource: DataSource,
+    private readonly msfaaNumberService: MSFAANumberService,
+  ) {
+    super(dataSource.getRepository(DisbursementSchedule));
+  }
+
+  /**
+   * Associates an MSFAA number to the disbursement(s) checking
+   * whatever is needed to create a new MSFAA or use an
+   * existing one instead.
+   * @param assessmentId assessment id of the workflow instance.
+   */
+  async associateMSFAANumber(assessmentId: number): Promise<void> {
+    const [firstDisbursement, secondDisbursement] = await this.repo.find({
+      select: {
+        id: true,
+        studentAssessment: {
+          id: true,
+          application: { id: true, student: { id: true } },
+          offering: { id: true, offeringIntensity: true, studyStartDate: true },
+        },
+        msfaaNumber: {
+          id: true,
+        },
+      },
+      relations: {
+        studentAssessment: { application: { student: true }, offering: true },
+        msfaaNumber: true,
+      },
+      where: {
+        studentAssessment: { id: assessmentId },
+      },
+    });
+
+    if (!firstDisbursement) {
+      throw new CustomNamedError(
+        "Disbursement not found or one of its associations is missing.",
+        DISBURSEMENT_NOT_FOUND,
+      );
+    }
+
+    if (firstDisbursement.msfaaNumber?.id) {
+      throw new CustomNamedError(
+        "MSFAA number is already associated.",
+        APPLICATION_MSFAA_ALREADY_ASSOCIATED,
+      );
+    }
+
+    let msfaaNumberId: number;
+
+    const studentId =
+      firstDisbursement.studentAssessment.application.student.id;
+    const offeringIntensity =
+      firstDisbursement.studentAssessment.offering.offeringIntensity;
+    const applicationId = firstDisbursement.studentAssessment.application.id;
+
+    // Checks if there is an MSFAA that could be considered valid.
+    const existingValidMSFAANumber =
+      await this.msfaaNumberService.getCurrentValidMSFAANumber(
+        studentId,
+        offeringIntensity,
+      );
+    if (existingValidMSFAANumber) {
+      // Reuse the MSFAA that is still valid and avoid creating a new one.
+      msfaaNumberId = existingValidMSFAANumber.id;
+    } else {
+      // Get previously completed and signed disbursement of an application for the student
+      // to determine if an existing MSFAA is still valid.
```

Comment:
I didn't follow the comment, if it is a valid MSAFAA, I  believe it should be part of line 78.

Reply:
I got the answer, I checked the function, `getCurrentValidMSFAANumber` it was looking ` qb.where("msfaaNumber.dateSigned is null") OR min valid date;`
![image](https://user-images.githubusercontent.com/77353155/228638741-fd4f9f96-4a90-4250-bfbf-f44da6c4f3f8.png)
![image](https://user-images.githubusercontent.com/77353155/228638942-90850190-022a-459d-8502-ff7cc2bc13a0.png)